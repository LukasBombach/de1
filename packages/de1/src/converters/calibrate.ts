import { Converter } from "@sblendid/sblendid";
import Parser from "../parser";

export interface Calibrate {
  writeKey: string;
  calCommand: number;
  calTarget: number;
  de1ReportedVal: number;
  measuredVal: number;
}

function decode(data: Buffer): Calibrate {
  return new Parser<Calibrate>(data)
    .int("writeKey", v => v.toString(16))
    .char("calCommand")
    .char("calTarget")
    .int("de1ReportedVal", v => Math.round(100 * (v / 65536)) / 100)
    .intSigned("measuredVal", v => Math.round(100 * (v / 65536)) / 100)
    .vars();
}

const converter: Converter<Calibrate> = {
  uuid: "a012",
  decode,
};

export default converter;
