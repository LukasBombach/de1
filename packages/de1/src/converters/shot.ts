import { Converter } from "@sblendid/sblendid";
import Parser from "../parser";

export interface Shot {
  timer: number;
  groupPressure: number;
  groupFlow: number;
  mixTemp: number;
  headTemp1: number;
  headTemp2: number;
  headTemp3: number;
  setMixTemp: number;
  setHeadTemp: number;
  setGroupPressure: number;
  setGroupFlow: number;
  frameNumber: number;
  steamTemp: number;
  time: number;
}

export const herz = 50;

function decode(data: Buffer): Shot {
  const time = Date.now();
  const vars = new Parser<Shot>(data)
    .short("timer", v => Math.round(100 * (v / (herz * 2))))
    .short("groupPressure", v => v / 4096)
    .short("groupFlow", v => v / 4096)
    .short("mixTemp", v => v / 256)
    .char("headTemp1")
    .char("headTemp2")
    .char("headTemp3")
    .short("setMixTemp", v => v / 256)
    .short("setHeadTemp", v => v / 256)
    .char("setGroupPressure", v => v / 16)
    .char("setGroupFlow", v => v / 16)
    .char("frameNumber")
    .char("steamTemp")
    .vars();
  return { ...vars, time };
}

const converter: Converter<Shot> = {
  uuid: "a00d",
  decode,
};

export default converter;
