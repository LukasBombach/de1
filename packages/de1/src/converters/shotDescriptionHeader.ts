import { ValidConverter } from ".";
import Parser from "../parser";
import Serializer from "../serializer";

export interface ShotDescriptionHeader {
  headerV: number;
  numberOfFrames: number;
  numberOfPreinfuseFrames: number;
  minimumPressure: number;
  maximumFlow: number;
}

function decode(data: Buffer): ShotDescriptionHeader {
  return new Parser<ShotDescriptionHeader>(data)
    .char("headerV")
    .char("numberOfFrames")
    .char("numberOfPreinfuseFrames")
    .char("minimumPressure", v => v / 16)
    .char("maximumFlow", v => v / 16)
    .vars();
}

function encode(header: ShotDescriptionHeader): Buffer {
  return new Serializer()
    .char(header.headerV)
    .char(header.numberOfFrames)
    .char(header.numberOfPreinfuseFrames)
    .char(header.minimumPressure * 16)
    .char(header.maximumFlow * 16).buffer;
}

function validate(header: Partial<ShotDescriptionHeader>): boolean {
  return Serializer.validate(header, {
    headerV: Number(),
    numberOfFrames: Number(),
    numberOfPreinfuseFrames: Number(),
    minimumPressure: Number(),
    maximumFlow: Number(),
  });
}

const converter: ValidConverter<ShotDescriptionHeader> = {
  uuid: "a00f",
  decode,
  encode,
  validate,
};

export default converter;
