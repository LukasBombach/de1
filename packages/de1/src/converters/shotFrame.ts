import { Converter } from "sblendid";
import Parser from "../parser";

export interface ShotFrame {
  frameToWrite: number;
  flag: number;
  setVal: number;
  temp: number;
  frameLen: number;
  triggerVal: number;
  maxVol: number;
}

function decode(data: Buffer): ShotFrame {
  return new Parser<ShotFrame>(data)
    .char("frameToWrite")
    .char("flag")
    .char("setVal", v => v / 16)
    .char("temp", v => v / 2)
    .char("frameLen", convertF817ToFloat)
    .char("triggerVal", v => v / 16)
    .short("maxVol", convertBottom10OfU10P0)
    .vars();
}

function convertF817ToFloat(value: number): number {
  const highBit = value & 128;
  return highBit === 0 ? value / 10 : value & 127;
}

function convertBottom10OfU10P0(value: number): number {
  return value & 1023;
}

const converter: Converter<ShotFrame> = {
  uuid: "a010",
  decode
};

export default converter;
