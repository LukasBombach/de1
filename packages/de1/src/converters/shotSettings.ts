import Parser from "../parser";
import Serializer from "../serializer";

export interface ShotSettings {
  steamSettings: number;
  targetSteamTemp: number;
  targetSteamLength: number;
  targetHotWaterTemp: number;
  targetHotWaterVol: number;
  targetHotWaterLength: number;
  targetEspressoVol: number;
  targetGroupTemp: number;
}

const converter: Converter<ShotSettings> = {
  name: "shotSettings",
  uuid: "a00b",
  encode,
  decode
};

function decode(data: Buffer): ShotSettings {
  return new Parser<ShotSettings>(data)
    .char("steamSettings")
    .char("targetSteamTemp")
    .char("targetSteamLength")
    .char("targetHotWaterTemp")
    .char("targetHotWaterVol")
    .char("targetHotWaterLength")
    .char("targetEspressoVol")
    .short("targetGroupTemp", v => v / 256)
    .vars();
}

function encode(shotSettings: ShotSettings): Buffer {
  return new Serializer()
    .char(shotSettings.steamSettings)
    .char(shotSettings.targetSteamTemp)
    .char(shotSettings.targetSteamLength)
    .char(shotSettings.targetHotWaterTemp)
    .char(shotSettings.targetHotWaterVol)
    .char(shotSettings.targetHotWaterLength)
    .char(shotSettings.targetEspressoVol)
    .short(shotSettings.targetGroupTemp)
    .buffer();
}

export default converter;
