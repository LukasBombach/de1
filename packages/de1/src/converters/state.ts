import { Converter } from "sblendid";
import Parser from "../parser";
import Serializer from "../serializer";

export type State =
  | "sleep"
  | "goingToSleep"
  | "idle"
  | "busy"
  | "espresso"
  | "steam"
  | "hotWater"
  | "shortCal"
  | "selfTest"
  | "longCal"
  | "descale"
  | "fatalError"
  | "init"
  | "noRequest"
  | "skipToNext"
  | "hotWaterRinse"
  | "steamRinse"
  | "refill"
  | "clean"
  | "inBootLoader"
  | "airPurge";

// type States = { [S in State]: number };
type States = Record<State, number>;

const converter: Converter<State> = {
  // name: "state",
  uuid: "a002",
  encode,
  decode
};

export const states: States = {
  sleep: 0x00,
  goingToSleep: 0x01,
  idle: 0x02,
  busy: 0x03,
  espresso: 0x04,
  steam: 0x05,
  hotWater: 0x06,
  shortCal: 0x07,
  selfTest: 0x08,
  longCal: 0x09,
  descale: 0x0a,
  fatalError: 0x0b,
  init: 0x0c,
  noRequest: 0x0d,
  skipToNext: 0x0e,
  hotWaterRinse: 0x0f,
  steamRinse: 0x10,
  refill: 0x11,
  clean: 0x12,
  inBootLoader: 0x13,
  airPurge: 0x14
};

function decode(data: Buffer): State {
  const value = parse(data);
  const state = getNameFromValue(value);
  if (!state) throw new Error(`Received unexpected state ${value}`);
  return state;
}

function encode(state: State): Buffer {
  const value = states[state];
  if (value === undefined) throw new Error(`Unknown state "${state}"`);
  return serialize(state);
}

function parse(data: Buffer): number {
  return new Parser<{ state: number }>(data).char("state").vars().state;
}

function serialize(state: State): Buffer {
  return new Serializer().char(states[state]).buffer();
}

function getNameFromValue(state: number): State {
  const values = Object.values(states);
  const keys = Object.keys(states);
  const valueIndex = values.indexOf(state);
  if (valueIndex === -1) throw new Error(`Invalid state value "${state}"`);
  return keys[valueIndex] as State;
}

export default converter;
