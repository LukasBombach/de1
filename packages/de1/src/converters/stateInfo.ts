import { Converter } from "sblendid";
import Parser from "../parser";
import { states, State, States } from "./state"; // todo this should be shared

export type SubState =
  | "ready"
  | "heating"
  | "finalHeating"
  | "stabilising"
  | "preinfusion"
  | "pouring"
  | "ending"
  | "refill";

export type SubStates = Record<SubState, number>;

export const subStates: SubStates = {
  ready: 0x00,
  heating: 0x01,
  finalHeating: 0x02,
  stabilising: 0x03,
  preinfusion: 0x04,
  pouring: 0x05,
  ending: 0x06,
  refill: 0x17
};

export interface StateInfo {
  state: State;
  substate: SubState;
}

function decode(data: Buffer): StateInfo {
  return new Parser<StateInfo>(data)
    .char("state", v => getKeyFromValue(states, v))
    .char("substate", v => getKeyFromValue(subStates, v))
    .vars();
}

function getKeyFromValue<Map extends States | SubStates>(
  map: Map,
  value: number
): Map extends States ? State : SubState {
  const values = Object.values(map);
  const keys = Object.keys(map);
  const valueIndex = values.indexOf(value);
  if (valueIndex === -1) throw new Error(`Invalid state value "${value}"`);
  return keys[valueIndex] as Map extends States ? State : SubState;
}

const converter: Converter<StateInfo> = {
  uuid: "a00e",
  decode
};

export default converter;
