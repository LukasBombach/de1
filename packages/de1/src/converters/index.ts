import { Converter } from "sblendid";

import state from "./state";
import water from "./water";
import versions from "./versions";
import shot from "./shot";
import shotSettings from "./shotSettings";
import stateInfo from "./stateInfo";

export type De1Converters = Converter<any>[];

const converters: De1Converters = [
  state,
  water,
  versions,
  shot,
  shotSettings,
  stateInfo
];

export default converters;
