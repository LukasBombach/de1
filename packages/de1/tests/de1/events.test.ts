import DE1 from "../../src/de1";
import converters, { Name } from "../../src/converters";

describe("de1 events", () => {
  const names = Object.keys(converters) as Name[];

  test.each(names)("%s emits the expected values", async name => {
    const de1 = new DE1();
    const listener = jest.fn();
    de1.on(name, listener);
    expect(listener.mock.calls).toMatchSnapshot();
  });
});
