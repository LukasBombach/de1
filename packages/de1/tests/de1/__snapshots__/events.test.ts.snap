// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`de1 events on(shot) emits the expected values 1`] = `
Array [
  Array [
    Object {
      "frameNumber": 3,
      "groupFlow": 0,
      "groupPressure": 0.010498046875,
      "headTemp": 24.933395385742188,
      "mixTemp": 25.03515625,
      "setGroupFlow": 0,
      "setGroupPressure": 0,
      "setHeadTemp": 89,
      "setMixTemp": 23.546875,
      "steamTemp": 33,
      "timer": 44742,
    },
  ],
]
`;

exports[`de1 events on(stateInfo) emits the expected values 1`] = `
Array [
  Array [
    Object {
      "state": "sleep",
      "substate": "ready",
    },
  ],
]
`;

exports[`de1 events on(water) emits the expected values 1`] = `
Array [
  Array [
    Object {
      "level": 5.1953125,
      "startFillLevel": 6,
    },
  ],
]
`;
